### PR TITLE
feat: Multi-domain, Introducing dataloader functionality

### PR DESCRIPTION
**Pull request by einrone**
## Description

This PR includes dataloader functionality that is being is used for multi-domain training. See
[Issue 919](https://github.com/ecmwf/anemoi-core/issues/919) for more in depth explanation.

## What problem does this change solve?

This PR may introduce a breaking change, where we have introduced a baseclass for multi-domain dataloader and multi-dataset

## What issue or task does this change relate to?
For more in-depth description see [Issue 919](https://github.com/ecmwf/anemoi-core/issues/919l)

##  Additional notes ##

Notice that this PR does include a test for multi-domain dataloader, but this is a small PR out of several PRs to enable multi-domain.Thus regarding this, we do not fully enable multi-domain yet, so the full solution i.e training cannot yet be tested.

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--1107.org.readthedocs.build/en/1107/

----
📚 Documentation preview 📚: https://anemoi-graphs--1107.org.readthedocs.build/en/1107/

----
📚 Documentation preview 📚: https://anemoi-models--1107.org.readthedocs.build/en/1107/

### CODE REVIEW

**Comment by einrone** [comment]
On ATS it was agreed no configuration will be provided in this PR, hence removing config changes and instantiation in datamodule.

File: training/tests/unit/data/test_multidomain.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from pytest_mock import MockFixture
+
+from anemoi.training.data.multidomain import MultiDomainDataset
+
+
+class TestMultiDomain:
+    """Test MultiDomainDataset instantiation and properties."""
+
+    @pytest.fixture
+    def multi_domain(self, mocker: MockFixture) -> MultiDomainDataset:
```
Comment (anaprietonem) [inline training/tests/unit/data/test_multidomain.py:22]: wouldn't it more representative to have 2 datasets with different frequencies,dates and resolution for the unit-tests of multi-domain?

File: training/src/anemoi/training/data/multidomain.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            worker_id : int
+                The ID of the current worker (0-indexed).
+
+        Returns
+        -------
+            None
+        """
+        self.worker_id = worker_id
+
+        for dataset in self.dataset_names:
```
Comment (anaprietonem) [inline training/src/anemoi/training/data/multidomain.py:134]: why do you have to loop over the datasets for the worker init in the multidomain dataset? We could at least have this inner loop as base_initialisation that get's reused for multi-ds and multi-domain

File: training/src/anemoi/training/data/multidomain.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            LOGGER.info(
+                ("Worker %d (%s, pid %d, base_seed %d, sanity rnd %f)"),
+                worker_id,
+                self.label,
+                os.getpid(),
+                base_seed,
+                sanity_rnd,
+            )
+
+    def get_sample(self, domain_name: str, index: int) -> dict[str, torch.Tensor]:
```
Comment (anaprietonem) [inline training/src/anemoi/training/data/multidomain.py:168]: @JPXKQX  you had some suggestions/ideas here to abstract this better with a sampler?

File: training/src/anemoi/training/data/anemoidataset.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+from torch.utils.data import IterableDataset
+
+from anemoi.models.distributed.balanced_partition import get_balanced_partition_sizes
+from anemoi.models.distributed.balanced_partition import get_partition_range
+from anemoi.training.data.data_reader import BaseAnemoiReader
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AnemoiDataset(IterableDataset, ABC):
```
Comment (anaprietonem) [inline training/src/anemoi/training/data/anemoidataset.py:32]: While discussing with @VeraChristina yesterday, anemoi-datasets now include units. Specially for multi-domain where we load many datasets in the same Datasets we should have a check that goes over the data_readers and inspect the variable_metadata for the units to ensure those are the same

Comment (anaprietonem) [reply training/src/anemoi/training/data/anemoidataset.py:32]: Similarly how do you that at the moment for the variable names in multi-domain @Sooph-knmi @einrone ?

File: training/src/anemoi/training/data/multidomain.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+from anemoi.training.utils.time_indices import normalize_time_indices
+from anemoi.training.utils.time_indices import offset_time_indices
+from training.src.anemoi.training.data.anemoidataset import AnemoiDataset
+
+LOGGER = logging.getLogger(__name__)
+
+# it may be that multidomain is not a good name for this, but it is what it is for now
+
+
+class MultiDomainDataset(AnemoiDataset):
```
Comment (anaprietonem) [inline training/src/anemoi/training/data/multidomain.py:33]: Can you add some documentation to explain what's a Multi-domain dataset?  We could extend https://anemoi.readthedocs.io/projects/training/en/latest/modules/data.html#multi-dataset

Comment (anaprietonem) [reply training/src/anemoi/training/data/multidomain.py:33]: Can a multi-domain dataset combine NativeGridDatasets and TrajectoryDatasets? or should we have some assert that the data_readers in multi-domain do not specify any trajectory?

**Review by anaprietonem — COMMENTED** [review]
Left a few comments, thanks for the nice work @einrone and @Sooph-knmi

File: training/src/anemoi/training/data/multidomain.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    def __iter__(self) -> Generator[dict[str, torch.Tensor], None, None]:
+        """Return an iterator that yields a tuple torch.Tensor and its corresponding domain name.
+
+        Returns
+        -------
+        tuple[torch.Tensor, str]
+            A tuple containing the tensor sample and its corresponding domain name
+        """
+        if self.shuffle:
```
Comment (anaprietonem) [inline training/src/anemoi/training/data/multidomain.py:196]: so in this case the shuffling is applied per dataset_reader since each can have it's own valid_date_indices, is that right?

File: training/tests/unit/data/test_multidomain.py (hunk 1 above)
Comment (einrone) [reply training/tests/unit/data/test_multidomain.py:22]: Yes that is true. However would you also like this test with an additional test which includes your suggestions?

File: training/src/anemoi/training/data/multidomain.py (hunk 2 above)
Comment (einrone) [reply training/src/anemoi/training/data/multidomain.py:134]: In the multi-domain dataloader you want to have control over how much data is partitioned per worker. That is the worker needs to have correct start end of time-indices of a given domain (we dont want to move into a time index that belongs to another domain) but also correct shard size for the given domain.  If that makes sense?

Comment (einrone) [reply training/src/anemoi/training/data/multidomain.py:134]: In the v1 implementation we did not implement this type of solution but it was quite unstable on different HPC, hence this solution fixed it.

File: training/src/anemoi/training/data/anemoidataset.py (hunk 4 above)
Comment (einrone) [reply training/src/anemoi/training/data/anemoidataset.py:32]: Yes, we do have a check for this. This check has been implemented in the refactored code in the DestinE repository but in the datamodule. If I understood you correctly, do you want me to include a check in the constructor or as a "private" method within the dataloader class?

File: training/tests/unit/data/test_multidomain.py (hunk 1 above)
Comment (Sooph-knmi) [reply training/tests/unit/data/test_multidomain.py:22]: Usually with multi-domain we would not have different frequencies (this is more something for multi-in/multi-out). But we do have different resolutions and dates. I can adapt the unit test to make it a bit more different.

File: training/src/anemoi/training/data/multidomain.py (hunk 5 above)
Comment (einrone) [reply training/src/anemoi/training/data/multidomain.py:33]: I mean in theory yes, I did implement this feature in the DestinE repository but it was an older version of the trajectory dataset. At that time this functionality was built into the single dataset class. However, I have not tested this functionality with multi-domain, have you @Sooph-knmi ?

File: training/src/anemoi/training/data/multidomain.py (hunk 6 above)
Comment (einrone) [reply training/src/anemoi/training/data/multidomain.py:196]: When shuffle=True, indices (dataset_reader for a given dataset) are shuffled independently for each dataset/domain first. Each domain is shuffled indices are then sliced to the worker’s assigned chunk. After that, the per-domain chunks are flattened into a single list of (domain, index) pairs. So this list includes all domains, because we want to sample domains randomly. This combined list is shuffled again, so samples from different domains are mixed before iteration. Finally, samples are yielded by looking up each (domain, index) pair.

**Comment by einrone** [comment]
> Left a few comments, thanks for the nice work @einrone and @Sooph-knmi

Hi, thank you for the reply and walking through the code. Really appreciate that, thanks. Let me know if there something which are a bit unclear or something that needs to be changed in order to fit into the current codebase.

File: training/src/anemoi/training/data/multidomain.py (hunk 6 above)
Comment (Sooph-knmi) [reply training/src/anemoi/training/data/multidomain.py:196]: Yes, but also, we need to have the same dataset across workers to avoid timeouts. So we shuffle each data reader shard per dataset before combining them, which enforces this property.

File: training/src/anemoi/training/data/multidomain.py (hunk 5 above)
Comment (Sooph-knmi) [reply training/src/anemoi/training/data/multidomain.py:33]: Hi @anaprietonem, as far as I can tell, TrajectoryDatasets are also not supported by multi-dataset? I am not so sure about how Anemoi deals with TrajectoryDatasets but maybe for now it is not a requirement to support it? In that case we can add an assertion as you mention.

Comment (Sooph-knmi) [reply training/src/anemoi/training/data/multidomain.py:33]: I took a quick look, it seems to be supported in the usable_indices selection. So I see no reasons why it would not be able to combine both datasets, as long as the correct indices are selected the samples can be shuffled and the correct sample can be passed to the model.

File: training/src/anemoi/training/data/multidomain.py (hunk 2 above)
Comment (anaprietonem) [reply training/src/anemoi/training/data/multidomain.py:134]: Thanks for the explanation, could we then try to at least reduce the code duplication and have the inner loop call a utility function from the base dataset?

File: training/src/anemoi/training/data/multidomain.py (hunk 5 above)
Comment (anaprietonem) [reply training/src/anemoi/training/data/multidomain.py:33]: Thanks for checking Sophie, indeed I am not trying to push for this to be supported mostly trying to ask if it makes sense to do it, and if not then suggest the code have an assert to prevent that.

File: training/src/anemoi/training/data/anemoidataset.py (hunk 4 above)
Comment (anaprietonem) [reply training/src/anemoi/training/data/anemoidataset.py:32]: Yes I'd go for a private method in the dataloader class. If useful this the work @VeraChristina  did for units https://github.com/ecmwf/anemoi-core/pull/1140/changes#diff-593510609c04c672c6651eff017a6c3ae8993ec36973d852e6ec4c52377d4081 so we should reuse something similar to https://github.com/ecmwf/anemoi-core/blob/a161913b3dd0bf23843a7b77feb875ef3f637597/training/src/anemoi/training/utils/variables_metadata.py

File: training/src/anemoi/training/data/multidomain.py (hunk 6 above)
Comment (anaprietonem) [reply training/src/anemoi/training/data/multidomain.py:196]: I think it would be good to document this better.

**Comment by einrone** [comment]
Resolved conflict with main and ran `test_multidomain.py` (2/2 passed 3 warnings) and `test_multidataset.py`  (5/5 passed 3 warnings). Warnings comes from deprecation warnings from anemoi regarding torch geometric and torch.jit.script

File: training/src/anemoi/training/data/multidomain.py (hunk 6 above)
Comment (einrone) [reply training/src/anemoi/training/data/multidomain.py:196]: Do you want that documentation in anemoi.readthedocs or briefly explained in a docstring?

**Comment by einrone** [comment]
@anaprietonem I added unit check in the dataloader and also created additional test, could you have a look if this is the implementation you were aiming for.

**Comment by anaprietonem** [comment]
> @anaprietonem I added unit check in the dataloader and also created additional test, could you have a look if this is the implementation you were aiming for.

Yes thanks a lot Aram for adding this! I think the only things missing area:
- Updating the docs
- Adding the assert/error catching if gridded and trajectory datasets are used
- Sampling mechanism we discussed with @JPXKQX  last week! 

feel to ping me for review again when those are added or if you have any questions!

File: training/src/anemoi/training/data/multidomain.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        }
+        # Normalize the date indices to use slices where possible, which can improve downstream indexing performance.
+        self.relative_date_indices = {
+            name: normalize_time_indices(indices) for name, indices in relative_date_indices.items()
+        }
+        LOGGER.info("valid date indices: %s", self.valid_date_indices)
+        self.n_samples_per_worker = {}  # overwrite base to empty dict
+        self.chunk_index_range = {}  # overwrite base to empty dict
+
+    def _check_datasets_units(self) -> None:
```
Comment (VeraChristina) [inline training/src/anemoi/training/data/multidomain.py:80]: This method doesn't seem to be called anywhere, unless I am missing something? Also, we noticed in other contexts that it can be useful to have an option to disable the checks, see e.g. https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/train/train.py#L337. We have added config options under the key `check_variables_compatibility` in several places in the training config (for the different checks), using the schema `CheckVariablesCompatibilitySchema`.

File: training/src/anemoi/training/data/datamodule.py [hunk 8]
```diff
@@ -136,6 +137,19 @@ def _get_dataset(
...
+                strategy,
+                data_readers=data_readers,
+                relative_date_indices=relative_date_indices,
+                shuffle=shuffle,
+                label=label,
+                epoch=self.epoch,
+                rollout=len(tuple(self.task.steps(label))),
+                **dataset_options,
+            )
+
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/datamodule.py]: Why not absorb this in the above, and make MultiDataset default instead of None? Or even break configs and demand for strategy to be set to `_target_.... MultiDataset` in the default case.

File: training/src/anemoi/training/data/multidomain.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        self.valid_date_indices = {
+            name: np.arange(len(anchors), dtype=np.int64) for name, anchors in self.anchors.items()
+        }
+        # Normalize the date indices to use slices where possible, which can improve downstream indexing performance.
+        self.relative_date_indices = {
+            name: normalize_time_indices(indices) for name, indices in relative_date_indices.items()
+        }
+
+    @cached_property
+    def frequency(self) -> dict[str, datetime.timedelta]:
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/multidomain.py]: Don't really see why this can't go and we simply inherit the `def frequency` from parent class?

File: training/src/anemoi/training/data/multidataset.py [hunk 10]
```diff
@@ -86,23 +86,23 @@ def __init__(
         if single_seq and multi_seq:
             msg = (
                 "Currently mixing single-sequence datasets (global time axis) with "
-                "Trajectory datasets (init x step axes) in the same MultiDataset is unsupported. "
+                f"Trajectory datasets (init x step axes) in the same {self.__class__.__name__} is unsupported. "
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/multidataset.py]:
```suggestion
                f"Trajectory datasets (init x step axes) in the same IterableDataset is unsupported. "
```

File: training/src/anemoi/training/data/multidataset.py [hunk 11]
```diff
@@ -86,23 +86,23 @@ def __init__(
...
                 f"Single-sequence: {single_seq}. Trajectory: {multi_seq}. "
             )
             raise ValueError(msg)
 
-        # Compute valid (sequence, position) anchors and a flat index over them
-        # that the shuffle/shard logic operates on.
+        self._set_date_indices(relative_date_indices)
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/multidataset.py:103]: when looking at the diff here, seems like many comments in the original code have dissappeared. Can they be added back, at the appropriate place?

File: training/src/anemoi/training/data/multidomain.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+        super().__init__(
+            data_readers=data_readers,
+            relative_date_indices=relative_date_indices,
+            shuffle=shuffle,
+            label=label,
+            epoch=epoch,
+            rollout=rollout,
+        )
+        self._check_datasets_units(**dict(check_variables_compatibility or {}))
+        LOGGER.info("valid date indices: %s", self.valid_date_indices)
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/multidomain.py]: Maybe this log message can be moved to the parent class? And maybe to debug? is this something we want to see in a typical log?

File: training/src/anemoi/training/data/multidomain.py (hunk 2 above)
Comment (dietervdb-meteo) [reply training/src/anemoi/training/data/multidomain.py]: Completely agree with Ana here, we can define a utility function 'set_samples_per_worker` in the parent class, and only override that. The rest of `per_worker_init` can then be directly inherited.

File: training/src/anemoi/training/data/multidomain.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+        self.n_samples_per_worker = {}
```
Comment (dietervdb-meteo) [inline training/src/anemoi/training/data/multidomain.py]: both of these can be moved to `per_worker_inits` (or even the sub-utility, see comment there), which is where these things get initialized anyways.

**Review by dietervdb-meteo — CHANGES_REQUESTED** [review]
Already some comments on streamlining the two DataSet classes even further.

**Comment by dietervdb-meteo** [comment]
PS I also unresolved a comment by Ana above, as I think she's right, and it wasn't fully taken into account.

**Comment by dietervdb-meteo** [comment]
I removed the breaking label, since we currently have fallbacks that make the old configs work. Personally I would be in favor of removing this extra code and make the PR (config) breaking. Lets discuss.

**Comment by einrone** [comment]
Quick rockachat or an actual meeting :) ?
